ExternalTest/rr_sim_tests/rr_sim_tests/utils/base_publisher.py

```diff
@@ -0,0 +1,143 @@
+#! /usr/bin/env python3
+# Copyright 2020-2023 Rapyuta Robotics Co., Ltd.
+
+import asyncio
+from threading import Thread
+
+# rclpy
+import rclpy
+
+# other ros
+from rclpy.executors import SingleThreadedExecutor
+from rclpy.node import Node
+from rclpy.qos import QoSDurabilityPolicy, QoSProfile, QoSReliabilityPolicy
+
+# Default number of publishing times, if <=0 will keep publishing until forced to stop
+PUBLISHER_DEFAULT_PUB_NUM = 1
+
+# Note: [PUBLISHER_DEFAULT_FREQ] here depends on [UE_FRAME_RATE & UE_DEFAULT_RTF] in such a way that the publishing 
+# should be timely enough for Sim to receive the msg
+PUBLISHER_DEFAULT_FREQ = 1.0 #Hz
+PUBLISHER_DEFAULT_QOS = QoSProfile(
+    depth=1,
+    durability=QoSDurabilityPolicy.RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL,
+    reliability=QoSReliabilityPolicy.RELIABLE,
+)
+
+class BasePublisher(Node):
+    def __init__(
+        self,
+        in_robot_namespace: str,
+        in_topic_name:str,
+        in_publish_num=PUBLISHER_DEFAULT_PUB_NUM,
+        in_publish_freq=PUBLISHER_DEFAULT_FREQ,
+        in_auto_publish=True
+    ):
+        super().__init__(node_name=f"{in_topic_name.replace('/','_')}_publisher", namespace=in_robot_namespace)
+
+        # https://github.com/ros2/rclcpp/blob/master/rclcpp/include/rclcpp/node_impl.hpp
+        # https://github.com/ros2/rclcpp/issues/1767
+        self._full_topic_name = None
+        if in_topic_name.startswith('/') or in_topic_name.startswith('~'):
```

Comment:
~ means private name space, i.e. it become nodename/topic name

Reply:
Does get_namespace include node name as well?
topic name should become ns/node_name/topic_name

Reply:
So it should be `get_effective_namespace()` in case of ~ I guess